pkg/serviceaccount/metadata/metadata.go

```diff
@@ -0,0 +1,157 @@
+	JWKSURI       string   `json:"jwks_uri"`                              // REQUIRED in OIDC; meaningful to relying parties.
+	ResponseTypes []string `json:"response_types_supported"`              // REQUIRED in OIDC
+	SubjectTypes  []string `json:"subject_types_supported"`               // REQUIRED in OIDC
+	SigningAlgs   []string `json:"id_token_signing_alg_values_supported"` // REQUIRED in OIDC
+
+}
+
+// NewServer creates a new IssuerMetadataServer.
+// The issuer is the OIDC issuer; keys are the keys that may be used to sign
+// KSA tokens.
+func NewServer(iss, jwksURI string, keys []interface{}) (*IssuerMetadataServer, error) {
+	// Metadata and keys are expected to only change across restarts at present,
+	// so we just marshal immediately and serve the cached JSON bytes.
+	keyset, errs := serviceaccount.PublicJWKSFromKeys(keys)
+	if errs != nil {
+		return nil, errs
+	}
+
+	keysetJSON, err := json.Marshal(keyset)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal service account issuer JWKS: %v", err)
+	}
+
+	metadata := issuerMetadata{
+		Issuer:        iss,
+		JWKSURI:       jwksURI,
+		ResponseTypes: []string{"id_token"}, // Kubernetes only produces ID tokens
+		SubjectTypes:  []string{"public"},   // https://openid.net/specs/openid-connect-core-1_0.html#SubjectIDTypes
+		SigningAlgs:   getAlgs(keyset),      // REQUIRED by OIDC
+	}
+
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal service account issuer metadata: %v", err)
+	}
+
+	return &IssuerMetadataServer{
+		metadataJSON: metadataJSON,
+		keysetJSON:   keysetJSON,
+	}, nil
+}
+
+// Install adds this server to the request router c.
+func (s *IssuerMetadataServer) Install(c *restful.Container) {
+	// Container.Add "will detect duplicate root paths and exit in that case",
+	// so we need a root for /.well-known/openid-configuration to avoid conflicts.
+	cfg := new(restful.WebService)
+	cfg.Path(OpenIDConfigPath).Route(
+		cfg.GET("").
+			To(fromStandard(s.serveConfiguration)).
+			Doc("get serviceaccount issuer OIDC configuration").
+			Operation("getServiceAccountIssuerMetadata"))
+	c.Add(cfg)
+	// ...and another one for the JWKS
+	jwks := new(restful.WebService)
+	jwks.Path(JWKSPath).Route(
+		jwks.GET("").
+			To(fromStandard(s.serveKeys)).
+			Doc("get serviceaccount issuer keys").
+			Operation("getServiceAccountIssuerKeys"))
+	c.Add(jwks)
+}
+
+// fromStandard provides compatibility between the standard (net/http) handler signature and the restful signature.
+func fromStandard(h http.HandlerFunc) restful.RouteFunction {
+	return func(req *restful.Request, resp *restful.Response) {
+		h(resp, req.Request)
+	}
+}
+
+func (s *IssuerMetadataServer) serveConfiguration(w http.ResponseWriter, req *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set("Cache-Control", cacheControlMaxAge)
```

Comment:
Can we also add 

- `Access-Control-Allow-Origin` set to `*` (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin)

This will help with browser to access via javascript XHR those resources. It's quite useful for clients interacting 

OIDC providers already serve those headers:

```
curl -I https://accounts.google.com/.well-known/openid-configuration
HTTP/2 200
accept-ranges: bytes
vary: Accept-Encoding
content-type: application/json
access-control-allow-origin: *
content-length: 1107
date: Tue, 14 Jan 2020 00:53:16 GMT
expires: Tue, 14 Jan 2020 01:53:16 GMT
last-modified: Thu, 09 Jan 2020 01:17:12 GMT
x-content-type-options: nosniff
server: sffe
x-xss-protection: 0
age: 3420
cache-control: public, max-age=3600
```

```
curl -I https://example.eu.auth0.com/.well-known/openid-configuration
HTTP/2 200
date: Tue, 14 Jan 2020 02:10:38 GMT
content-type: application/json; charset=utf-8
content-length: 1398
server: nginx
ot-tracer-spanid: 5b3ed08823bd7a6f
ot-tracer-traceid: 6dba3ee01092f6f4
ot-tracer-sampled: true
x-auth0-requestid: 304112cba01e3e4be355
access-control-allow-origin: *
access-control-allow-credentials: false
etag: W/"576-ZXy0TSdluCp2BwBvxYpz5CiCOek"
cache-control: private, no-store, no-cache, must-revalidate, post-check=0, pre-check=0, no-transform
strict-transport-security: max-age=15724800
x-robots-tag: noindex, nofollow, nosnippet, noarchive
```

Reply:
That is already an option on the kube-apiserver:
```
$ kube-apiserver --help | grep cors
      --cors-allowed-origins strings                 List of allowed origins for CORS, comma separated.  An allowed origin can be a regular expression to support subdomain matching. If this list is empty CORS will not be enabled.
```

Reply:
I forgot that this flag applies to all endpoints.